std/assembly/string.ts

```diff
@@ -393,7 +394,127 @@ export class String {
     return result;
   }
 
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replace(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var index = this.indexOf(search);
+    if (~index) {
+      let rlen = replacement.length;
+      len = len - slen;
+      let olen = len + rlen;
+      if (olen) {
+        let result = allocateUnsafe(olen);
+        copyUnsafe(result, 0, this, 0, index);
+        copyUnsafe(result, index, replacement, 0, rlen);
+        copyUnsafe(result, rlen + index, this, index + slen, len - index);
+        return result;
+      }
+    }
+    return this;
+  }
+
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replaceAll(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var rlen = replacement.length;
+    if (!slen) {
+      // Handle special case when we should insert replacement between each char
+      // TODO handle surrogate pairs as well
```

Comment:
Can you give an example what's not working without this?

Reply:
i'm not sure this relate to us because most of string operations in JS now unicode aware. But it seems other languages care about this. I actually thinking of add special boolean optinal param to methods like `split`, `reverse` and etc which turn on utf safe mode. But need more investigation. I could remove this comment currently